rpki-validator/src/main/java/net/ripe/rpki/validator3/domain/validation/CertificateTreeValidationService.java

```diff
@@ -401,31 +402,35 @@ private CertificateAuthorityValidationResult validateManifestEntry(
             X509Crl crl,
             Map.Entry<String, byte[]> entry
     ) {
-        Key key = Key.of(entry.getValue());
-        URI location = manifestUri.resolve(entry.getKey());
-        ValidationResult validations = newValidationResult(location);
+        ValidationResult validations = newValidationResult(manifestUri);
         CertificateAuthorityValidationResult result = CertificateAuthorityValidationResult.of(validations);
 
-        Optional<RpkiObject> object = storage.readTx(tx -> rpkiObjects.findBySha256(tx, entry.getValue()));
-        validations.rejectIfFalse(object.isPresent(), VALIDATOR_MANIFEST_ENTRY_FOUND, manifestUri.toASCIIString());
+        String entryFilename = entry.getKey();
+        byte[] entryHash = entry.getValue();
+        URI entryLocation = manifestUri.resolve(entryFilename);
+
+        Optional<RpkiObject> object = storage.readTx(tx -> rpkiObjects.findBySha256(tx, entryHash));
+        validations.rejectIfFalse(object.isPresent(), VALIDATOR_MANIFEST_ENTRY_FOUND, entryFilename);
         if (validations.hasFailureForCurrentLocation()) {
             return result;
         }
 
         RpkiObject rpkiObject = object.get();
-        boolean hashMatches = Arrays.equals(rpkiObject.getSha256(), entry.getValue());
-        validations.rejectIfFalse(hashMatches, VALIDATOR_MANIFEST_ENTRY_HASH_MATCHES, entry.getKey());
+        boolean hashMatches = Arrays.equals(Sha256.hash(rpkiObject.getEncoded()), entryHash);
+        validations.rejectIfFalse(hashMatches, VALIDATOR_MANIFEST_ENTRY_HASH_MATCHES, entryFilename);
         if (validations.hasFailureForCurrentLocation()) {
             return result;
         }
 
         SortedSet<String> locations = storage.readTx(tx -> rpkiObjects.getLocations(tx, rpkiObject.key()));
-        validations.rejectIfFalse(locations.contains(location.toASCIIString()), VALIDATOR_REPOSITORY_NOT_AT_EXPECTED_LOCATION, location.toASCIIString());
+        validations.rejectIfFalse(locations.contains(entryLocation.toASCIIString()), VALIDATOR_REPOSITORY_NOT_AT_EXPECTED_LOCATION, entryLocation.toASCIIString());
```

Comment:
Looks ok,  just trying to understand better. If object is present based on check on line 413, can we still have a case where `entryLocation` is not on list of `locations`?

Reply:
In theory the object could be published in some other repository. Since we perform all our lookups by SHA-256 hash we do  not necessarily know where the object was located. This check ensures it was located at the expected location.